x-pack/legacy/plugins/actions/server/plugin.ts

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Hapi from 'hapi';
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+import { ActionsConfigType, Services } from './types';
+import { ActionExecutor, TaskRunnerFactory } from './lib';
+import { ActionsClient } from './actions_client';
+import { ActionTypeRegistry } from './action_type_registry';
+import { ExecuteOptions } from './create_execute_function';
+import { createExecuteFunction } from './create_execute_function';
+import { registerBuiltInActionTypes } from './builtin_action_types';
+import { IClusterClient, KibanaRequest, Logger } from '../../../../../src/core/server';
+import { getActionsConfigurationUtilities } from './actions_config';
+import {
+  ActionsPluginInitializerContext,
+  ActionsCoreSetup,
+  ActionsCoreStart,
+  ActionsPluginsSetup,
+  ActionsPluginsStart,
+} from './shim';
+import {
+  createActionRoute,
+  deleteActionRoute,
+  findActionRoute,
+  getActionRoute,
+  updateActionRoute,
+  listActionTypesRoute,
+  getExecuteActionRoute,
+} from './routes';
+
+export interface PluginSetupContract {
+  registerType: ActionTypeRegistry['register'];
+}
+
+export interface PluginStartContract {
+  listTypes: ActionTypeRegistry['list'];
+  execute(options: ExecuteOptions): Promise<void>;
+  getActionsClientWithRequest(request: Hapi.Request): ActionsClient;
+}
+
+export class Plugin {
+  private readonly config$: Observable<ActionsConfigType>;
+  private readonly logger: Logger;
+  private serverBasePath?: string;
+  private adminClient?: IClusterClient;
+  private taskRunnerFactory?: TaskRunnerFactory;
+  private actionTypeRegistry?: ActionTypeRegistry;
+  private actionExecutor?: ActionExecutor;
+
+  constructor(initializerContext: ActionsPluginInitializerContext) {
+    this.logger = initializerContext.logger.get('plugins', 'alerting');
+    this.config$ = initializerContext.config.create();
+  }
+
+  public async setup(
+    core: ActionsCoreSetup,
+    plugins: ActionsPluginsSetup
+  ): Promise<PluginSetupContract> {
+    const config = await this.config$.pipe(first()).toPromise();
+    this.adminClient = await core.elasticsearch.adminClient$.pipe(first()).toPromise();
```

Comment:
I've been wondering if this is kosher or not.  Is there an expectation that the we should be doing the `await ...toPromise()` dance somewhere, where the `adminClient` is actually used?

We'll find out! (I'm doing the same as this in the event log plugin :-) )

Reply:
Yeah I was thinking the same, like what if we shouldn't just be grabbing `first()`.